File: src/main/java/me/testaccount666/serversystem/commands/executables/inventorysee/online/CommandInventorySee.java
package me.testaccount666.serversystem.commands.executables.inventorysee.online;

import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.inventorysee.offline.CommandOfflineInventorySee;
import me.testaccount666.serversystem.commands.executables.inventorysee.utils.InventorySeeUtils;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.ConsoleUser;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.command.Command;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;

import java.util.HashMap;
import java.util.Map;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "inventorysee", variants = "offlineinventorysee", tabCompleter = TabCompleterInventorySee.class)
public class CommandInventorySee extends AbstractServerSystemCommand {
    public final Map<Player, Inventory> inventoryCache = new HashMap<>();
    public final CommandOfflineInventorySee offlineInventorySee;

    public CommandInventorySee() {
        offlineInventorySee = new CommandOfflineInventorySee(this);
    }

    @Override
    public void execute(User sender, Command command, String label, String... arguments) {
        if (command.getName().toLowerCase().startsWith("offline")) {
            offlineInventorySee.execute(sender, command, label, arguments);
            return;
        }

        if (!command.getName().equalsIgnoreCase("inventorysee")) return;
        processInventorySee(sender, label, arguments);
    }

    public void processInventorySee(User sender, String label, String... arguments) {
        if (!checkBasePermission(sender, "InventorySee.Use")) return;

        if (sender instanceof ConsoleUser) {
            general("NotPlayer", sender).build();
            return;
        }

        if (arguments.length < 1) {
            general("InvalidArguments", sender).syntax(getSyntaxPath(null)).label(label).build();
            return;
        }

        var targetUserOptional = getTargetUser(sender, arguments[0]);
        if (targetUserOptional.isEmpty()) {
            general("PlayerNotFound", sender).target(arguments[0]).build();
            return;
        }

        var targetUser = targetUserOptional.get();
        if (targetUser == sender) {
            command("InventorySee.CannotSeeSelf", sender).build();
            return;
        }

        var ownerPlayer = targetUser.getPlayer();
        var viewerPlayer = sender.getPlayer();

        var customInventory = inventoryCache.computeIfAbsent(ownerPlayer, this::createAndInitializeInventory);
        viewerPlayer.openInventory(customInventory);
    }

    private Inventory createAndInitializeInventory(Player ownerPlayer) {
        var newInventory = Bukkit.createInventory(ownerPlayer, 54, "${ownerPlayer.getName()}'s Inventory");
        refreshInventoryContents(ownerPlayer, newInventory);
        return newInventory;
    }

    protected void refreshInventoryContents(Player owner, Inventory displayInventory) {
        copyPlayerInventory(owner, displayInventory);
        addSectionDecorators(displayInventory);
    }

    private void copyPlayerInventory(Player owner, Inventory displayInventory) {
        // main inventory and hotbar (slots 0-39)
        for (var slot = 0; slot < 40; slot++) displayInventory.setItem(slot, owner.getInventory().getItem(slot));

        displayInventory.setItem(40, owner.getInventory().getItemInOffHand());
        displayInventory.setItem(41, owner.getOpenInventory().getCursor());
    }

    public void addSectionDecorators(Inventory displayInventory) {
        InventorySeeUtils.addSectionDecorators(displayInventory, false);
    }

    public void placeFilledMarkers(Inventory inventory, Material material, String displayName, int startSlot, int endSlot) {
        InventorySeeUtils.placeFilledMarkers(inventory, material, displayName, startSlot, endSlot);
    }

    protected void applyChangesToOwner(Player owner, Inventory displayInventory) {
        for (var slot = 0; slot < 40; slot++) owner.getInventory().setItem(slot, displayInventory.getItem(slot));

        owner.getInventory().setItemInOffHand(displayInventory.getItem(40));
        owner.setItemOnCursor(displayInventory.getItem(41));
        owner.updateInventory();
        dropModifiedItems(owner, displayInventory, 42, 45);
    }

    private void dropModifiedItems(Player owner, Inventory displayInventory, int fromSlot, int toSlot) {
        for (var slot = fromSlot; slot < toSlot; slot++) {
            var item = displayInventory.getItem(slot);
            if (item == null) continue;

            owner.getWorld()
                    .dropItem(owner.getEyeLocation().add(0, -0.33, 0), item.clone())
                    .setVelocity(owner.getLocation().getDirection().multiply(0.35));
            item.setAmount(0);
        }
    }

    @Override
    public String getSyntaxPath(Command command) {
        return "InventorySee";
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        if (command.getName().toLowerCase().startsWith("offline")) return offlineInventorySee.hasCommandAccess(player, command);

        return PermissionManager.hasCommandPermission(player, "InventorySee.Use", false);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/inventorysee/online/CommandInventorySee.kt
package me.testaccount666.serversystem.commands.executables.inventorysee.online

import me.testaccount666.paperktx.extensions.*
import me.testaccount666.paperktx.extensions.ComponentExtensions.asComponent
import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.commands.executables.inventorysee.offline.CommandOfflineInventorySee
import me.testaccount666.serversystem.commands.executables.inventorysee.utils.InventorySeeUtils
import me.testaccount666.serversystem.extensions.commandMsg
import me.testaccount666.serversystem.extensions.generalMsg
import me.testaccount666.serversystem.userdata.User
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.command.Command
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory

@ServerSystemCommand("inventorysee", ["offlineinventorysee"], TabCompleterInventorySee::class)
class CommandInventorySee : AbstractServerSystemCommand() {
    val inventoryCache = HashMap<Player, Inventory>()
    val offlineInventorySee = CommandOfflineInventorySee(this)
    override fun minRequiredArguments(command: Command): Int {
        if (isOffline(command)) return offlineInventorySee.minRequiredArguments(command)
        return 1
    }

    override fun getUsagePermission(command: Command): String {
        if (isOffline(command)) return offlineInventorySee.getUsagePermission(command)
        return "InventorySee.Use"
    }

    override fun getSyntaxPath(command: Command?): String {
        command ?: return "InventorySee"
        if (isOffline(command)) return offlineInventorySee.getSyntaxPath(command)
        return "InventorySee"
    }

    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        if (command.name.startsWith("offline", true)) {
            offlineInventorySee.execute(commandSender, command, label, *arguments)
            return
        }

        if (!command.name.equals("inventorysee", true)) return
        processInventorySee(commandSender, label, *arguments)
    }

    fun processInventorySee(sender: User, label: String, vararg arguments: String) {
        if (!isPlayer(sender)) return

        val targetUser = getTargetUser(sender, arguments = arrayOf(arguments[0])) ?: run {
            sender.generalMsg("PlayerNotFound") { target(arguments[0]) }
            return
        }

        if (targetUser === sender) {
            sender.commandMsg("InventorySee.CannotSeeSelf")
            return
        }

        val ownerPlayer = targetUser.getPlayer()!!
        val viewerPlayer = sender.getPlayer()!!

        inventoryCache.computeIfAbsent(ownerPlayer, ::createAndInitializeInventory).also(viewerPlayer::openInventory)
    }

    private fun createAndInitializeInventory(ownerPlayer: Player): Inventory {
        return Bukkit.createInventory(ownerPlayer, 54, "${ownerPlayer.name}'s Inventory".asComponent()).apply {
            refreshInventoryContents(ownerPlayer, this)
        }
    }

    fun refreshInventoryContents(owner: Player, displayInventory: Inventory) {
        copyPlayerInventory(owner, displayInventory)
        addSectionDecorators(displayInventory)
    }

    private fun copyPlayerInventory(owner: Player, displayInventory: Inventory) {
        // main inventory and hotbar (slots 0-39)
        for (slot in 0..39) displayInventory[slot] = owner.inventory[slot]

        displayInventory[40] = owner.inventory.itemInOffHand
        displayInventory[41] = owner.itemOnCursor
    }

    fun addSectionDecorators(displayInventory: Inventory) {
        InventorySeeUtils.addSectionDecorators(displayInventory, false)
    }

    fun placeFilledMarkers(inventory: Inventory, material: Material, displayName: String, startSlot: Int, endSlot: Int) {
        InventorySeeUtils.placeFilledMarkers(inventory, material, displayName, startSlot, endSlot)
    }

    fun applyChangesToOwner(owner: Player, displayInventory: Inventory) {
        for (slot in 0..39) owner.inventory[slot] = displayInventory[slot]

        owner.inventory.setItemInOffHand(displayInventory[40])
        owner.setItemOnCursor(displayInventory[41])
        owner.updateInventory()
        dropModifiedItems(owner, displayInventory, 42, 45)
    }

    private fun dropModifiedItems(owner: Player, displayInventory: Inventory, fromSlot: Int, toSlot: Int) {
        displayInventory[fromSlot..<toSlot].filterNot { it.isAir() }.forEach {
            owner.world.dropItem(owner.eyeLocation.add(0.0, -0.33, 0.0), it.clone()).velocity = owner.location.direction.multiply(0.35)
            it.amount = 0
        }
    }

    private fun isOffline(command: Command) = command.name.startsWith("offline", true)
}
